### PR TITLE
Add network Address class

### PR DESCRIPTION
**Pull request by roby2014**
blocked by #1259 

# Description

Add a network `Address` class, as proposed in the issue.

## Checklist

- [x] Self-review changes.
- [x] Evaluate impact on the documentation.
- [x] Ensure test coverage.
- [x] Write new samples.
- [x] Add entry to the changelog's unreleased section.

### CODE REVIEW

File: core/tests/net/address.cpp [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+#include <cubos/core/net/address.hpp>
+
+TEST_CASE("address")
+{
+    using cubos::core::net::Address;
+
+    SUBCASE("from and to string")
+    {
+        auto addr = Address::from("127.0.0.1");
```
Comment (RiscadoA) [inline core/tests/net/address.cpp:11]: Does `Address::from("localhost")` return `127.0.0.1`? If so, we should cover it

**Review by RiscadoA — CHANGES_REQUESTED** [review]
Thanks for working on this! Nice to see this moving forward :eyes: maybe our next game will have online multiplayer. Added some nitpicks.

File: core/src/net/address.cpp [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+const Address Address::LocalHost(127, 0, 0, 1);
+
+CUBOS_REFLECT_IMPL(cubos::core::net::Address)
+{
+    using cubos::core::reflection::StringConversionTrait;
+    using cubos::core::reflection::Type;
+
+    return Type::create("Address").with(
+        StringConversionTrait{[](const void* instance) { return static_cast<const Address*>(instance)->toString(); },
+                              [](void*, const std::string& string) { return Address::from(string).has_value(); }});
```
[the remark this answers is not in the document]
Comment (roby2014) [reply core/src/net/address.cpp:25]: Right. I have a question, should we write unit tests for these `Address` traits?

[the remark this answers is not in the document]
Comment (RiscadoA) [reply core/src/net/address.cpp:25]: Good idea yeah, that would be best.

File: core/tests/net/address.cpp [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        REQUIRE(addr.has_value());
+        REQUIRE(addr.value().toString() == "127.0.0.1");
+
+        REQUIRE(!Address::from("abc123").has_value());
+    }
+
```
[the remark this answers is not in the document]
Comment (roby2014) [reply core/tests/net/address.cpp:17]: That wont work, `from` user `inet_pton` which accepts only IPv4 addresses in dotted-decimal notation.

File: core/tests/net/address.cpp (hunk 1 above)
Comment (roby2014) [reply core/tests/net/address.cpp:11]: Same problem as above conversation,

File: core/tests/net/address.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (RiscadoA) [reply core/tests/net/address.cpp:17]: Hm, but there are alternatives which allow it right? Being able to resolve a domain name is essential for accessing servers hosted by the developers themselves for example.

File: core/include/cubos/core/ecs/system/schedule.hpp [hunk 4]
```diff
@@ -94,19 +94,19 @@ namespace cubos::core::ecs
         /// @param conditionId Condition identifier.
         /// @param repeatId Repeat node to become part of, if any.
         /// @return Repeat node identifier, or nothing if the given repeat node is not a repeat node.
-        Opt<NodeId> repeat(ConditionId conditionId, Opt<NodeId> repeatId = {});
+        memory::Opt<NodeId> repeat(ConditionId conditionId, memory::Opt<NodeId> repeatId = {});
```
[the remark this answers is not in the document]
Comment (roby2014) [reply core/include/cubos/core/ecs/system/schedule.hpp:97]: @RiscadoA why would this error get catched here but not in #1259 :|

[the remark this answers is not in the document]
Comment (RiscadoA) [reply core/include/cubos/core/ecs/system/schedule.hpp:97]: Scary, better fix it there ig. I have no idea

**Review by EdSwordsmith — APPROVED** [review]

**Review by joaomanita — COMMENTED** [review]
It would be nice to have 100% line coverage or close to that in the test cases.

**Comment by roby2014** [comment]
> It would be nice to have 100% line coverage or close to that in the test cases.

Im 99% sure that codecov is borked.

**Review by joaomanita — APPROVED** [review]

**Comment by RiscadoA** [comment]
> > It would be nice to have 100% line coverage or close to that in the test cases.
> 
> Im 99% sure that codecov is borked. It should be > 90%

You didn't add the test in the CMakeLists.txt, so its not being compiled & run

**Comment by roby2014** [comment]
Thanks @RiscadoA , nice catch.

File: core/include/cubos/core/net/address.hpp [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+        ~Address();
+
+        /// @brief Constructs a new IPv4 address from four 8-bit octets (`a.b.c.d`).
+        /// @param a First octet.
+        /// @param b Second octet.
+        /// @param c Third octet.
+        /// @param d Fourth octet.
+        static Address fromIPv4(uint8_t a, uint8_t b, uint8_t c, uint8_t d);
+
+        /// @brief Constructs a new IPv6 address from four 32-bit octets (`a.b.c.d`).
```
Comment (RiscadoA) [inline core/include/cubos/core/net/address.hpp:43]:
```suggestion
        /// @brief Constructs a new IPv6 address from four 32-bit integers (`a.b.c.d`).
```
No longer octets

File: core/include/cubos/core/net/address.hpp [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+        /// @brief 127.0.0.1
+        static const Address LocalHost;
+
+        /// @brief Constructs a empty address (IPv4 by default).
+        Address();
+
+        /// @brief Deconstructs by cleaning necessary resources.
+        ~Address();
+
+        /// @brief Constructs a new IPv4 address from four 8-bit octets (`a.b.c.d`).
```
Comment (RiscadoA) [inline core/include/cubos/core/net/address.hpp:36]:
```suggestion
        /// @brief Constructs a new IPv4 address from four octets (`a.b.c.d`).
```
Redundant

**Review by RiscadoA — APPROVED** [review]
LGTM! Thanks for working on this

File: core/src/net/address.cpp [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+    , mType(IPType::IPv4)
+{
+    init();
+}
+
+Address::Address(uint32_t address)
+    : mAddr{.v4Addr = address}
+    , mType(IPType::IPv4)
+{
+    init();
```
Comment (RiscadoA) [inline core/src/net/address.cpp:79]: Maybe its better to call `init` only when you actually use the functions, and clean up immediately after, as you do in `from`.

**Pull request merged**